pkg/ordermanager/ordermanager.go

```diff
@@ -0,0 +1,410 @@
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Error creating new order: %v", err)
+		return ticket{}, err
+	}
+
+	eg := &errgroup.Group{}
+	for _, zurl := range o.AuthzURLs {
+		z, err := om.Client.GetAuthorization(ctx, zurl)
+		if err != nil {
+			return ticket{}, err
+		}
+		// Find the HTTP01 challenge (all we support)
+		chal, err := getHTTP01(z.Challenges)
+		if err != nil {
+			return ticket{}, err
+		}
+		resp, err := om.Client.HTTP01ChallengeResponse(chal.Token)
+		if err != nil {
+			return ticket{}, err
+		}
+		path := om.Client.HTTP01ChallengePath(chal.Token)
+		om.Challenger.RegisterChallenge(path, resp)
+
+		eg.Go(func() error {
+			defer om.Challenger.UnregisterChallenge(path)
+
+			// TODO(mattmoor): Wait until we have successfully probed the
+			// challenge ourselves before "Accepting" to get positive hand-off
+			// to the routing layer that things have been successfully plumbed.
+			// something something wait.Until()
+			time.Sleep(2 * time.Second)
+
+			if _, err := om.Client.Accept(ctx, chal); err != nil {
+				return err
+			}
+			if _, err := om.Client.WaitAuthorization(ctx, z.URI); err != nil {
+				return err
+			}
+			return nil
+		})
+	}
+
+	go func() {
+		if err := eg.Wait(); err != nil {
+			logging.FromContext(ctx).Errorf("Encountered an error waiting for challenges: %v.", err)
+			om.setError(ctx, domains, err)
+		} else if _, err := om.Client.WaitOrder(ctx, o.URI); err != nil {
+			logging.FromContext(ctx).Errorf("Encountered an error waiting for order: %v.", err)
+			om.setError(ctx, domains, err)
+		} else {
+			logging.FromContext(ctx).Infof("Order %q has completed without error.", o.URI)
+		}
+
+		// The order is ready for fulfillment (one way or another)!
+		om.Callback(owner)
+	}()
+
+	logging.FromContext(ctx).Infof("Order %q has been initiated.", o.URI)
+	t := ticket{
+		uri:   o.URI,
+		owner: owner,
+	}
+	func() {
+		om.Lock()
+		defer om.Unlock()
+		om.inflight[asKey(domains)] = t
+	}()
+	return t, nil
+}
+
+func (om *impl) completeOrder(ctx context.Context, domains []string, t ticket) (*tls.Certificate, error) {
+	cert, err := t.GetCertificate(ctx, om.Client, domains)
+	if err != nil {
+		return nil, err
+	}
+
+	om.Lock()
+	defer om.Unlock()
+
+	delete(om.inflight, asKey(domains))
+	return cert, nil
+}
+
+func (om *impl) cancelOrder(ctx context.Context, domains []string) {
+	om.Lock()
+	defer om.Unlock()
+
+	delete(om.inflight, asKey(domains))
+}
+
+func (om *impl) setError(ctx context.Context, domains []string, err error) {
+	om.Lock()
+	defer om.Unlock()
+
+	t, ok := om.inflight[asKey(domains)]
+	if !ok {
+		return
+	}
+	t.err = err
+	om.inflight[asKey(domains)] = t
+}
+
+func (t *ticket) GetStatus(ctx context.Context, client *acme.Client) (string, error) {
+	o, err := client.GetOrder(ctx, t.uri)
+	if err != nil {
+		return "", err
+	}
+	return o.Status, nil
+}
+
+func (t *ticket) GetError(ctx context.Context, client *acme.Client) (error, error) {
```

Comment:
name the output vars to distinguish between the two types of errors?

Reply:
done